### ISSUE TITLE
Store app settings outside of the ddoc

### ISSUE THREAD

**Issue by garethbowen** [medic/medic-webapp#4541]
For historical reasons we store the configuration in the app_settings property of the main ddoc. These historical reasons are no longer relevant so we are free to store the configuration in a different doc. This is good because it...

1. Simplifies the upgrade code because the new ddoc can be uploaded without worry about overwriting settings. This also helps deployment in a developer environment and unblocks #4536.
2. Greatly reduces the amount of data users have to download when settings are changed.

**Comment by garethbowen** [comment]
Two PRs to review please @SCdF 

Most of the code is directly replacing getting and updating settings from the ddoc with the settings doc. The tricky bits are the migrations. If upgrading using gardener (if this is supported?) gardener will leave the settings on the ddoc and the api ddoc-extraction code will update the settings doc. If upgrading using Horti the second PR will update the settings doc before the ddoc is uploaded.

**Comment by SCdF** [comment]
Back to you @garethbowen

**Comment by garethbowen** [comment]
@SCdF Thanks for that. I've added some questions to the webapp PR.

Also, how do we go about releasing Horti? Should I leave this to you?

**Comment by SCdF** [comment]
You can do it when you're ready or I can. Just increment the minor and `npm publish`. Someone other than me should probably have proven credentials about thi---

OH WAIT. Actually your horti code is arguably not right, because it does not maintain backwards compatibility with versions prior to this change. Apologies for not catching that earlier.

So probably the simplest way of doing this is to just have horti copy settings in the ddoc if they exist, otherwise do nothing, and then have the migration deal with the moving in api.

Or we can just give up and clean out everything in the horti staging server before 3.0.0, but I'd rather we didn't since we've managed to maintain backwards compat so far and we should probably get into the habit of thinking this way.

**Comment by garethbowen** [comment]
Ok... so the new upgrade workflow is:

1. Horti hasn't changed - when upgrading the app_settings are copied from the old medic ddoc to the new medic ddoc then uploaded.
2. API v3.0 begins startup process
3. DDoc extraction runs which pulls out the secondary ddocs including medic-client. Because it no longer copies the app_settings into medic-client we detect that the medic-client ddoc has changed and update it with no settings. It has also changed in other ways (views, etc) but this means the settings are no longer on the client ddoc.
4. The `config` module is inited which queries the `settings` service which looks for the `settings` doc, finds none, and fetches it from the `ddoc.app_settings` for backwards compatibility.
5. The new `extract-settings` migration runs before (almost) all other migrations which gets the ddoc app_settings, and overwrites the settings in the `settings` doc, then deletes the ddoc app_settings and saves the ddoc.
6. DDoc extraction notices the update and checks if the secondary ddocs need to be updated - they don't.
7. The `config` module notices the update and reloads the in memory config.
8. API starts up and starts serving requests.
9. Later on Horti does another upgrade but there are no app_settings on the ddoc so nothing to copy.

**Comment by garethbowen** [comment]
Horti hasn't changed so no release necessary.

**Comment by newtewt** [comment]
Was able to validate this on beta.dev that https://beta.dev.medicmobile.org/medic/_design/medic/ no longer contains the settings and are at https://beta.dev.medicmobile.org/medic/settings. Also compared https://beta-old.dev.medicmobile.org/medic/_design/medic/ against https://beta.dev.medicmobile.org/medic/_design/medic/ and locally went from 2.16.0 to 3.0.0-beta1 to see the move occur.
